src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -59,9 +65,10 @@ BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
     }
 
     bid.set(m_last_append_offset, 1, m_chunk_id);
+
+    [[maybe_unused]] auto cur_cp = hs()->cp_mgr().cp_guard();
```

Comment:
You may need to be move the m_last_append_offset and m_freeable_blks to the current CPContext and change the CPContext when you do `on_switchover_cp`. Current change might not protect from the race between cp modification and flush.

Reply:
Changing CPContext is already done `on_swtichover_cp`. 
That (moving offset/freeable_nblks to CPContext) is something I was saying yesterday, if we move to CPContext, we will need ThreadVector/RCU for different blk_allocator.  
But why, isn't this good enough for us make sure either all of them are in current cp or next cp. Nothing will be in between?

Reply:
We don't need threadvector/RCU because CPContext* on CP is already an RCU. You need to assign all your members (need a containing structure) (offset/freeable_nblks) into the CPContext* inside the CP. Now you assign this structure to CPContext*. When you switchover CP, you copy these containing structure and attach to next CP.

I put in my suggestion here:

```
class AppendBlkAllocator {

struct AllocContext {
   offset;
   freeable_nblks;
};

on_switchover_cp(old_cp, new_cp) {
      new_cp->cp_ctx->alloc_context = make_unique< AllocContext >();
      // copy offset, freeable_nblks from old_cp
      *new_cp->cp_ctx->alloc_context  = *old_cp->context->alloc_context;
}

alloc_blk() {
    cp = CPGuard();
    cp->cp_ctx->alloc_context->...(all current logic)
}

flush_cp() {
   // Take data from cp->cp_ctx and flush its details.
}
};
```